File: Lecture4Exercise/allchr_pca_project.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib.backends.backend_pdf import PdfPages

# Load data
pca = pd.read_csv("chr20_pca_results.eigenvec", sep='\t').rename(columns={'#FID': 'FID'})
panel = pd.read_csv("integrated_call_samples_v3.20130502.ALL.panel", sep='\t')
df_orig = pd.merge(pca, panel, left_on='IID', right_on='sample')

# Load projected score (PLINK 2 .sscore)
projected = pd.read_csv("related_projection.sscore", sep='\t')
# Map score columns to PC names
projected = projected.rename(columns={'IID': 'IID', 'SCORE1_AVG': 'PC1', 'SCORE2_AVG': 'PC2'})

with PdfPages('allchr_pca_project.pdf') as pdf:
    plt.figure(figsize=(10, 7))
    
    # Background: Unrelated individuals
    sns.scatterplot(data=df_orig, x='PC1', y='PC2', hue='super_pop', alpha=0.3, palette='Set2', edgecolor=None)
    
    # Foreground: The projected individual
    plt.scatter(projected['PC1'], projected['PC2'], color='black', marker='X', s=200, label='Projected Indiv')
    
    plt.title("Projection of Filtered Individual onto 1000G PCA space")
    plt.legend(bbox_to_anchor=(1.05, 1), loc='upper left')
    plt.tight_layout()
    pdf.savefig()
    plt.close()

File: Lecture4Exercise/allchr_pca_project.R
library(ggplot2)

# Load original PCA and Panel
pca_orig <- read.table("allchr.EUR.biallelicsnps_unrelated_pruned_pca.eigenvec")
colnames(pca_orig)=c("FID", "IID",paste0("PC",c(1:10)))
panel <- read.table("integrated_call_samples_v3.20130502.ALL.panel",header=T)
df_orig <- merge(pca_orig, panel, by.x = "IID", by.y = "sample")

# Load Projected Individual (from .sscore file)
# Note: PLINK 2 score output columns are usually SCORE1_AVG, etc.
projected <- read.table("related_projection.sscore")
colnames(projected)=c("FID","IID","SCORE1_AVG", "SCORE2_AVG",paste0("PC",c(1:10)))

pdf("allchr.EUR.biallelicsnps_pca_project.pdf", width = 10, height = 8)

ggplot() +
  # Plot the background (original unrelated individuals)
  geom_point(data = df_orig, aes(x = PC1, y = PC2, color = super_pop), alpha = 0.3) +
  # Plot the projected individual as a large distinct star/point
  geom_point(data = projected, aes(x = PC1, y = PC2), color = "black", shape = 18, size = 5) +
  geom_text(data = projected, aes(x = PC1, y = PC2, label = "Related Individual"), vjust = -1.5) +
  theme_minimal() +
  labs(title = "PCA Projection: Related Individual onto 1000G Space",
       x = "PC1", y = "PC2")

dev.off()
